druid-shell/src/windows/mod.rs

```diff
@@ -268,300 +294,378 @@ impl WndProc for MyWndProc {
+                if let Ok(mut s) = self.state.try_borrow_mut() {
+                    let s = s.as_mut().unwrap();
+                    let width = LOWORD(lparam as u32) as u32;
+                    let height = HIWORD(lparam as u32) as u32;
+                    let mut ctx = WinCtxImpl {
+                        handle: &self.handle.borrow(),
+                    };
+                    s.handler.size(width, height, &mut ctx);
+                    let use_hwnd = if let Some(ref dcomp_state) = s.dcomp_state {
+                        dcomp_state.sizing
+                    } else {
+                        true
+                    };
+                    if use_hwnd {
+                        if let Some(ref mut rt) = s.render_target {
+                            if let Some(hrt) = cast_to_hwnd(rt) {
+                                let width = LOWORD(lparam as u32) as u32;
+                                let height = HIWORD(lparam as u32) as u32;
+                                let size = SizeU(D2D1_SIZE_U { width, height });
+                                let _ = hrt.resize(size);
+                            }
                         }
-                        ValidateRect(hwnd, null_mut());
+                        InvalidateRect(hwnd, null_mut(), FALSE);
                     } else {
-                        println!("ResizeBuffers failed: 0x{:x}", res);
+                        let res;
+                        {
+                            s.render_target = None;
+                            res = (*s.dcomp_state.as_mut().unwrap().swap_chain).ResizeBuffers(
+                                0,
+                                width,
+                                height,
+                                DXGI_FORMAT_UNKNOWN,
+                                0,
+                            );
+                        }
+                        if SUCCEEDED(res) {
+                            s.rebuild_render_target(&self.d2d_factory);
+                            s.render(&self.d2d_factory, &self.dwrite_factory, &self.handle);
+                            if let Some(ref mut dcomp_state) = s.dcomp_state {
+                                (*dcomp_state.swap_chain).Present(0, 0);
+                                let _ = dcomp_state.dcomp_device.commit();
+                            }
+                            ValidateRect(hwnd, null_mut());
+                        } else {
+                            println!("ResizeBuffers failed: 0x{:x}", res);
+                        }
                     }
+                } else {
+                    self.log_dropped_msg(hwnd, msg, wparam, lparam);
                 }
                 Some(0)
             },
             WM_COMMAND => {
-                self.handler.command(LOWORD(wparam as u32) as u32);
+                if let Ok(mut s) = self.state.try_borrow_mut() {
+                    let s = s.as_mut().unwrap();
+                    let mut ctx = WinCtxImpl {
```

Comment:
the fact that this gets repeated so often makes me think it might be worth having a `WinCtxImpl::new` fn, save us a couple lines in each match arm.

Reply:
That'd be nice, but I tried a couple variants and wasn't able to make it work. That function would need to hold a reference to `&mut self`, and calling the handler method also requires `&mut self`.

There could well be a way to reduce the duplication, I just don't see it.